### PR TITLE
Python binding improvements - mark 2

### PR DESCRIPTION
**Pull request by ikelos**
Hiya, I just tried building the python bindings for linux and found pull request #14.  I've tried to update it to the support the latest version, but I have to admit my C code isn't all that great.

It compiles just fine, but `LcCreate` fails and I've no idea how to add the structure needed to `LcCreateEx` to see what's actually causing the problem.  On the same system I can run leechcore with the same device parameter and it reads just fine.

I'm happy to help test and develop improvements, but my C skills are pretty basic.  Anything that can help drive this forward a bit would be appreciated.  It feels like so much work has already been done, there's just a little bit extra needed to get linux supported too...  5:)

### CODE REVIEW

**Comment by ufrisk** [comment]
Thank You.

I've thought about the possibly putting it up on pypi. But I'm not sure what the best way forward here is so I kinda looked the other way for a long time. But maybe you have some ideas; I think it would be great if one could make PCILeech more accessible for other projects to integrate with.

About the pull request(s) themselves: I'm not able to accept them as-is due to copyright reasons. I mention this in the project readme. Accepting it would bind me towards the GPL and I would effectively lose the ability to re-license under separate licenses. The solution for this is to set up a CLA. It looks like I'll have to, within the next few months, set up a small side business around this, and after that I would be able to look into the CLA part. The business is quite costly though so I've been putting it off for long with this just being a hobby project with hardly any revenues.

Windows:
* PCILeech with the FPGA require a FTDIChip FT601 .dll which may only be OK to re-distribute if it accompanies a hardware device as per: https://ftdichip.com/driver-licence-terms-details/ I cannot guarantee this.
* Building binaries on Windows as part of the installation process is a big no-no; but binary distribution and together with the MemProcFS should be fine here anyway.

Linux:
* PCILeech with the FPGA requires a driver here as well; either a separate kernel-driver or a libusb driver (user-mode) that I've put as a separate .so in this repo: https://github.com/ufrisk/LeechCore-plugins since it's being contributed to me under GPL and for reasons outlined above I have this as an optional plugin. I believe this may be the reason why it fails to connect, or if you're not running as root or with invalid device permissions in /dev/. It should be possible to bundle the libusb driver in a GPL'ed linux pypi package though.
* I'm guessing building from source on linux is a must - i.e. not distribute the .so binaries themselves?

The LcCreateEx is mostly used for receiving "error" info when connecting to live Hyper-V machines - i.e. to allow the API caller to pop up a VM-select prompt to the user. It's not relevant for Linux. Yet anyway. I have plans for a linux compatible authentication mechanism towards the remote Windows agent I have.

Also, I've been putting this API off a long time; most people use the more feature-rich Windows-only MemProcFS for API integration into their projects. I haven't seen any interest in this area from the Linux side; but now when there is I'll have to look into it somehow.

What would be easiest, short term, for me would be to make it build on Linux and include the .so in the current LeechCore distribution. You wouldn't get the convenience of pip though. But would this be helpful for your integration use case?

**Comment by ikelos** [comment]
I agree that packaging for pypi containing binary components is likely difficult.  I guess look at things like distorm/capstone for how to do that?

The README doesn't really explain *why* you don't accept contributions, but that's fine.  The changes aren't a significant work and I fully trust you could and possibly will make them on your own, but if you don't have time to work on python support this may help motivate others to try it out and develop it.  If you get a CLA in place I'd probably sign it, I'm not really in it for profit I just want a cool tool that everyone can use at the end of it.  5;)

I'd really like to see example code with LeechCorePyc so that the compiled library can be tested with known working code and possibly also to give the API a feel and see if it can be made easier (you might want to stick with the LeechCore API as closely as possible, but if you'd done the work of making a python API it might be worth deviating in order to make it more pythonic?).  I don't have any figures to back up the claim, but I suspect that a strong interpreted interface like python will get people playing with the technology more than having to wrestle with C to then play with it.

Thanks for the pointers, I'll check out the optional driver.  The fact that it works on the system installed pcileech suggest that's not the problem, but it might be a linking issue or something, given the python module seems to be a complete rebuild of the library.  I was looking at a CFFI interface but all the defines for Windows/Linux made that a little tricky so I haven't gotten too far with it (particularly once I found leechcorepyc).

Errr, I don't think build from source on linux is necessarily a must, but offering people the possibility will likely go down well (I use a source-based distribution, so I much appreciate having the option).  It will also mean that the build process will get stress tested and you'll end up with quite a robust system if you have the time and energy to fix the bugs that'll get filed.  5;)

Ah, ok, I thought it might be for general errors like "can't find library" or "USB device not plugged in" or similar.  It would be useful to expose such troubleshooting responses from the call, but I understand if that's tricky.  I haven't gone anywhere near the agent yet, so I don't have a great deal to comment on it.

Getting it to build (and work) on linux with instructions (so pointers about where the FPGA libraries need to be etc) would be awesome!  Packaging it as a binary might be nice to make it slightly more convenient but it's certainly not critical to me personally.

In short, any help you can provide would be appreciated and if you can help get past the hurdles so I can start building on it, then I can make more use of it and contribute/suggest things that might make stuff easier.  I'm also happy to wait for a CLA before expecting merges...  5;)  Thanks Ulf!

**Comment by ikelos** [comment]
Not sure if this is the right place for discussion, happy to move it elsewhere if there's somewhere better.

I managed to get it going and found the following:
* Even changing the python library to build against the library rather than compile a copy into the module, it still always looks in the directory that the executable is run from to find plugins.  This means `/usr/bin` because it's python that's running and python lives at `/usr/bin/python`.
* `LEECHCOREPYC_Open` works fine, but rather than returning you an object containing the handle (that could then have methods called on it, like read, etc), it opens a global handle, meaning you can only have one open at a time, and calls like `LEECHCOREPYC_Read` only ever act on the one global open handle.
* `LEECHCOREPYC_Read(offset, length)` didn't seem to work (I tried multiple offsets and multiple lengths, I only ever got an empty bytes object back)?
* `LEECHCOREPYC_ReadScatter([list of offsets])` worked fine, and returned me the pages that had been requested in the list.  This is also what leads me to believe that `LEECHCOREPYC_Read` isn't working, because the same offset put in `ReadScatter` worked fine.

So things that would make this much better/easier for *linux* users to use would be:
* Move the library to `libleechcore-x.y.z.so` following the standard library numbering scheme (https://www.gnu.org/software/libtool/manual/html_node/Updating-version-info.html).  It's very similar to semantic versioning if that helps?  It needs to start with lib so that all the linkers and so on can find it (`-lleechcore` in gcc wil look for `libleechcore` when searching for the library).
* Load the plugins from the library's/a fixed location, rather than the executable using the library.

So things that would make this much better/easier for *python* users to use would be:
* Have `LEECHCOREPYC_Open` return the handle and pass that in to other `LEECHCOREPYC` calls.
* Drop the `LEECHCOREPYC` at the start of the function names, it's a bit cumbersome to put in, particularly since the module is already called `leechcorepyc`.  5;)
* Refactor the `LEECHCOREPYC` calls, so instead `Open` creates an object and the other functions turn into methods on that object, so that you could end up writing something like:
```
lc = leechcorepyc.open('fpga://', '')
data = lc.read(0x1000, 0x2000)
lc.close()
```

I know there's a lot there and it's absolutely a niche part of a niche project that probably doesn't consume all your time in the first place, so no problem on if/when it gets done, but I figured I'd let you know what I think would help and then at least it's out there.  I can at the moment pull data back using what's there, it's fully usable as is, so thanks for that!!!  5:D  If you are interested in making any of the changes, or get a CLA that you're happy with then I'm happy to help test, etc where I can...  5:)

**Comment by ufrisk** [comment]
Apologies for the somewhat slow answer; I've been busy with another thing this weekend.

This sounds great to me overall; we should really discuss more; just drop me a note at pcileech@frizk.net or catch me on Twitter or Discord UlfFrisk#5780

I'm quite aware that DMA and memory analysis is a nieche product; that's why I've always had a full time job and this "just" being my hobby project. I spend way too much time overall across my projects. Even though unrealistic to make a living out of it I have received some minor contributions mainly via the github sponsors - which is nice since github is doubling the amount.

I have also finally gotten permission, from work, to start a small side business around this, which I need to do to keep the tax man happy...

I don't believe it to be realistic to make much more out of this other than cover some expenses, but in the past I've received almost zero source code contributions, and after talking to some legal people I just don't want to throw away that very remote possibility of making something real out of this by accepting contributions without a CLA. Especially very insignificant contributions. I also accept modules contributions (pypykatz being one). I need to set up my side business first though; I'm guessing around a month for this.

---

As you already probably noticed the LeechCore Python library isn't really used. The few people that use my Python integrations are using the MemProcFS Python library - but that's Windows only unfortunately and it's not likely to change.

I started the Python library quite some time ago; since then I've upgraded the LeechCore to 2.0 to support a multi-handle/thread/device architecture - but the LeechCore Python library isn't upgraded. I just put very little time into that. And I'm not very good with Python.

Step one if doing a refresh around this would probably be to rename the functions, dump the LEECHCOREPYC to something more friendly - alternatively add a Python layer on-top of it just as I did with the VmmPy API for MemProcFS. https://github.com/ufrisk/MemProcFS/blob/master/files/vmmpy.py and also add support for multi-handle/multi-device.

Also, partly related, I should add some way to the Linux version of the LeechCore to communicate with the Windows-only LeechAgent (currently it's Windows RPC with Active Directory Kerberos Auth only) - that would allow things like connecting to remote DMA and also remote live Hyper-V machines even from Linux.

But the LeechCore is only ever going to be handling physical memory - the MemProcFS is for virtual memory.

Anyway, I'd love to get some help around this when I get the paperwork sorted out, but catch me on Twitter or Discord :)

**Comment by ufrisk** [comment]
I'm closing this for now. The plan is to release an updated version of this python binding within the next 1-2 months; and possibly on python pip as well as discussed separately.

**Comment by ikelos** [comment]
Just noticed a couple more things, and this seems like a reasonable place to keep track of them, even though the issue's closed:

* It would be really handy to know how big the accessible memory is, I've done this in the existing implementation by adding:
`PyDict_SetItemString_DECREF(pyDict, "maxAddr", PyLong_FromLong(cfg.paMax));`.  If it's being refactored into a python object, then having this accessible in the object would be massively useful.
* The `LEECHCOREPYC_Read` function currently fails because it uses cbRead which is set to 0 at the start of the call and then never changed.  It starts returning results when `cbRead` is changed to `cb` (although I haven't chekced if the `LcRead` call will change `cb` to say how many bytes were actually read).
* Even after changing that `LEECHCOREPYC_Read` always reads from offset 0, because the argument doesn't get converted properly.  I don't know why, parsing the args tuple looks ok, the format string looks ok, but `pa` always comes back as 0 (even if it's initialized to something different).
